Add the ability to preempt jobs via deletion

bhroam opened the pull request:
#### Describe Bug or Feature
Add ability to preempt jobs via deletion

#### Describe Your Change
There is a new option to preempt_order (D) which signifies to preempt via deletion.  
The req_preempt_jobs.c file got refactored.  Now each preemption method is achieved by creating an internal batch request.  This walks through the same code path as the scheduler did when it preempted jobs itself (e.g. qsig, qdel, qhold, etc).

Previous to this change, the preempt jobs preq was stored as a nested batch request in the new preq.  This has changed.  I now store the preempt jobs preq on the job itself.  This is due to when a job is fully deleted and purged, the preq is not available.  The job is available in all cases where we need to reply to the preempt jobs preq.

#### Link to Design Doc
https://pbspro.atlassian.net/wiki/spaces/PD/pages/1197735939/Preemption+via+deletion


#### Attach Test and Valgrind Logs/Output
[preemption.log](https://github.com/PBSPro/pbspro/files/3209901/preemption.log)


<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->



<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

arungrover commented on src/server/req_holdjob.c:
```diff
@@ -145,25 +145,37 @@ req_holdjob(struct batch_request *preq)
 	if ((jt != IS_ARRAY_NO) && (jt != IS_ARRAY_ArrayJob)) {
+		/* Need to check if the subjob is being preempted */
+		pjob = find_job(preq->rq_ind.rq_hold.rq_orig.rq_objname);
+		if(pjob != NULL && pjob->ji_pmt_preq != NULL)
```
Please add a space after 'if'

arungrover commented on src/server/req_holdjob.c:
```diff
@@ -145,25 +145,37 @@ req_holdjob(struct batch_request *preq)
-	if ((rc=get_hold(&preq->rq_ind.rq_hold.rq_orig.rq_attr, &pset)) != 0) {
+	if ((rc = get_hold(&preq->rq_ind.rq_hold.rq_orig.rq_attr, &pset)) != 0) {
+		if (pjob->ji_pmt_preq != NULL)
+			reply_preempt_jobs_request(PBSE_IVALREQ, PREEMPT_METHOD_CHECKPOINT, pjob);
```
Why don't we use rc here instead of PBSE_IVALREQ?

arungrover commented on src/server/req_delete.c:
```diff
@@ -1270,7 +1285,7 @@ post_delete_mom1(struct work_task *pwt)
 			}
 			/* 2nd try, use SIGTERM */
-			rc = issue_signal(pjob, sigt, post_delete_mom1, preq_clt, NULL);
+			rc = issue_signal(pjob, sigt, post_delete_mom1, preq_clt);
```
Don't we need to reply to preempt req on line 1307 as well? when mom rejects the request stating the job was not found.

arungrover commented on src/server/req_rerun.c:
```diff
@@ -120,8 +122,8 @@ post_rerun(struct work_task *pwt)
 		}
-	} else if (preq->rq_nest)
-		reply_preempt_jobs_request(PBSE_NONE, 3, preq);
+	} else if (pjob->ji_pmt_preq != NULL)
```
We should check for pjob to be !NULL here

bhroam commented on src/include/job.h:
```diff
@@ -516,7 +516,8 @@ struct job {
 	int		ji_mom_prot;	/* rpp or tcp */
 	struct batch_request *ji_rerun_preq;	/* outstanding rerun request */
-	int		ji_licneed;	/* # of cpu licenses needed by job */
+	struct batch_request *ji_pmt_preq;		/* outstanding preempt job request for deleting jobs */
```
I fixed this by moving ji_pmt_preq into the server only portion of the job structure.

bhroam commented on src/server/req_delete.c:
```diff
@@ -421,6 +421,8 @@ req_deletejob(struct batch_request *preq)
 
 		if ((i == JOB_STATE_EXITING) && (forcedel == 0)) {
+			if (parent->ji_pmt_preq != NULL) 
+				reply_preempt_jobs_request(PBSE_BADSTATE, PREEMPT_METHOD_DELETE, parent);
```
I didn't really need to handle this case since you'll never get here.  If the job is not in the running state, you don't do a delete job request.  You handle the preemption right away.  I did as you suggested since I could though.  I did a find_job()

bhroam commented on src/server/req_delete.c:
```diff
@@ -421,6 +421,8 @@ req_deletejob(struct batch_request *preq)
+				reply_preempt_jobs_request(PBSE_BADSTATE, PREEMPT_METHOD_DELETE, parent);
 			req_reject(PBSE_BADSTATE, 0, preq);
 			return;
 		} else if (i == JOB_STATE_EXPIRED) {
```
If the subjob is expired, we won't find it in req_preeemptjobs() and the delete request won't be sent.

bhroam replied:
There is a race, but we'll catch that case in req_preemptjobs().  We first check if we can find the job or the job is not in the running state.  We will reply back for that job immediately either successfully or fail it depending on its state.  In this case where the job expired, we'll reply back as it was successfully deleted.

bhroam commented on src/server/req_holdjob.c:
```diff
@@ -390,13 +407,14 @@ post_hold(struct work_task *pwt)
 	preq->rq_conn = preq->rq_orgconn;	/* restore client socket */
-
+	
+	pjob = find_job(preq->rq_ind.rq_hold.rq_orig.rq_objname);
```
OK, I moved some code around so the NULL check happens first.  If pjob is NULL, we can't reply to the preempt request, but that shouldn't matter.  There should be no way we can get there if pjob is NULL without already replying.  If the job ended normally, we'll reply as if it was deleted.  If a qdel -Wforce happened, we'll reply there.  We should be safe.

bhroam commented on src/server/req_preemptjob.c:
```diff
@@ -72,6 +71,199 @@ extern struct server server;
 extern time_t time_now;
 
+static int preempt_total = 0;
+static int preempt_index = 0;
```
I didn't think about multi-sched.  The current code used static variables, so I made them global.  I'm going to have to think about how to fix preemption for multi-sched.  Good catch.

prakashcv13 replied:
should we make these part of the pbs_sched object, maybe a struct?

bhroam commented on src/server/req_preemptjob.c:
```diff
@@ -72,6 +71,199 @@ extern struct server server;
+
+/**
+ * @brief create a local batch_request for a suspend request
+ * @param[in] job_id - the job to create the request for
```
Good catch.  I needed to add it for all of the functions to create batch requests.

bhroam commented on src/server/req_preemptjob.c:
```diff
@@ -72,6 +71,199 @@ extern struct server server;
+	if (newreq == NULL || hold_svrattrl == NULL) {
+		if (newreq != NULL)
+			free_br(newreq);
+		if (hold_svrattrl != NULL)
```
Since there are two, it could be an else if.  I'd rather leave it alone because if in the future we add a third variable we need to allocate for, we'll need to make it a series of ifs then.  If you really want it an else if, I'll do it.

bhroam commented on src/server/req_preemptjob.c:
```diff
@@ -126,112 +357,54 @@ req_preemptjobs(struct batch_request *preq)
-				break;
+		if (pjob->preempt_order[0].order[pjob->preempt_order_index] == PREEMPT_METHOD_CHECKPOINT)
+			clear_preempt_hold(pjob);
+		if (issue_preempt_request((int)pjob->preempt_order[0].order[pjob->preempt_order_index + 1], pjob, preq)) {
```
The reason I did it this way is control will fall down to where I check if we're done and send the preempt reply.  This might be the final job, and we need to send the ack back to the scheduler.  If we made an else and returned from there, I'd have to move the check to reply in there as well as the bottom of the function.  I think it is good as it is.  Let me know what you think.

arungrover replied:
You are right, I misread the code :)

bhroam commented on src/server/req_holdjob.c:
```diff
@@ -145,25 +145,37 @@ req_holdjob(struct batch_request *preq)
 		return;
 	if ((jt != IS_ARRAY_NO) && (jt != IS_ARRAY_ArrayJob)) {
+		/* Need to check if the subjob is being preempted */
+		pjob = find_job(preq->rq_ind.rq_hold.rq_orig.rq_objname);
```
chk_job_request() will find the the parent array if it is a subjob.  I need to find the subjob itself.  I'll make the comment more clear.

bhroam commented on src/server/req_preemptjob.c:
```diff
@@ -85,34 +277,73 @@ extern time_t time_now;
 	for (i = 0; i < count; i++) {
 		ppj = &(preq->rq_ind.rq_preempt.ppj_list[i]);
 		pjob = find_job(ppj->job_id);
+		/* The job is already out of the way.  This must of happened after the scheduler
```
Well this comment led me down an interesting rabbit hole.  20+ years ago when I was taught to type, two spaces was the norm after a period.  I talked to @agurban and she said that one or two is fine and she uses two.  She also pointed me at the Chicago Manual of Style as the definitive place to look.  They said the CMoS says one space.  Life changes when you move from a monospaced typewriter to a computer.  I guess I have old habits and will have to start changing.  Since most people who work on PBS started typing on a computer, they will use one space.  I'll be consistent and change it.

bhroam commented on src/server/req_holdjob.c:
```diff
@@ -405,26 +423,27 @@ post_hold(struct work_task *pwt)
-		if (preq->rq_nest)
-			reply_preempt_jobs_request(PBSE_SYSTEM, 0, preq);
+		if (pjob->ji_pmt_preq != NULL)
+			reply_preempt_jobs_request(PBSE_SYSTEM, PREEMPT_METHOD_CHECKPOINT, pjob);
```
This code was incorrect.  It was in a block of if (pjob == NULL).  This would have dereferenced pjob and crashed.  I removed it.  We shouldn't get here since we catch the case where the job is not there in req_preemptjob()

bhroam commented:
@prakashcv13 @arungrover I have addressed your comments.  Please take another look.
New testlog: [preemption.log](https://github.com/PBSPro/pbspro/files/3214654/preemption.log)

arungrover commented on src/server/req_preemptjob.c:
```diff
@@ -72,6 +71,207 @@ extern struct server server;
+	old_hold = pjob->ji_wattr[(int)JOB_ATR_hold].at_val.at_long;
+	job_attr_def[(int)JOB_ATR_hold].at_set(&pjob->ji_wattr[(int)JOB_ATR_hold],
+					       &temphold, DECR);
+	log_event(PBSEVENT_JOB, PBS_EVENTCLASS_JOB, LOG_INFO, pjob->ji_qs.ji_jobid, log_buffer);
```
We didn't write anything to log_buffer before logging

arungrover commented on src/server/req_preemptjob.c:
```diff
@@ -72,6 +71,207 @@ extern struct server server;
+
+	if (old_hold != pjob->ji_wattr[(int)JOB_ATR_hold].at_val.at_long) {
+		pjob->ji_modified = 1; /* indicates attributes changed    */
+		log_event(PBSEVENT_JOB, PBS_EVENTCLASS_JOB, LOG_INFO, pjob->ji_qs.ji_jobid, log_buffer);
```
Same comment as above

arungrover commented on src/server/req_preemptjob.c:
```diff
@@ -85,34 +285,73 @@ extern time_t time_now;
+			sprintf(preempt_jobs_list[preempt_index].job_id, "%s", ppj->job_id);
+			strcpy(preempt_jobs_list[preempt_index].order, "D");
+			preempt_index++;
+			job_preempt_fail(preq, ppj->job_id);
```
why do we need to fail it? You are already setting the job_id and order in the lines above

arungrover replied:
@bhroam how about this comment?

arungrover commented on src/server/req_preemptjob.c:
```diff
@@ -85,34 +285,73 @@ extern time_t time_now;
+					preempt_index++;
+					break;
+				case JOB_STATE_EXPIRED:
+				case JOB_STATE_FINISHED:
```
How about also adding JOB_STATE_MOVED, in case the job moved to peer complex

bhroam commented:
Thanks for your comments @arungrover.  I have addressed them.

prakashcv13 commented on src/server/req_delete.c:
```diff
@@ -697,14 +708,13 @@ req_deletejob2(struct batch_request *preq, job *pjob)
 			pjob->ji_qs.ji_substate = JOB_SUBSTATE_RUNNING;
 			log_event(PBSEVENT_JOB, PBS_EVENTCLASS_JOB, LOG_INFO,
-				pjob->ji_qs.ji_jobid, "deleting instead of reruning");
+				  pjob->ji_qs.ji_jobid, "deleting instead of reruning");
```
shall we make it "rerunning"

bhroam commented on src/server/node_manager.c:
```diff
@@ -2715,6 +2715,13 @@ discard_job(job *pjob, char *txt, int noack)
+
+	if (pjob->ji_pmt_preq != NULL)
+		reply_preempt_jobs_request(PBSE_NONE, PREEMPT_METHOD_DELETE, pjob);
+
```
We won't call this function twice because it NULLs ji_pmt_preq.  If we come across another time we could call the function, ji_pmt_preq will be NULL, so we won't call it.

bhroam commented on src/server/req_delete.c:
```diff
@@ -663,7 +667,8 @@ req_deletejob2(struct batch_request *preq, job *pjob)
 				} else {
-
+					if (pjob->ji_pmt_preq != NULL)
+						reply_preempt_jobs_request(PBSE_SYSTEM, PREEMPT_METHOD_DELETE, pjob);
```
reply_preempt_jobs_request() will only reply to the preq that the scheduler sent.  The preq in this function is the locally created preq that was created in req_preemptjobs().  If we didn't reject it, it'd never be freed.

bhroam commented:
@prakashcv13 about moving the global variable to the sched object.  We either need to move them to the sched object, or at least have an array of them.  In any case, since this was an existing bug, I filed a ticket for it.  I am not planning on fixing it in this PR.

bhroam commented:
I've WIP'd this PR.  I was writing a couple more tests and ran into some bugs and a crash.  I am tracking it down.

bhroam commented:
OK, I removed the WIP tag.  The crash turned out to be a typo in the code I modified (not the reviewed code).  I wrote 2 more tests which test qsub -r n and qsub -c n.  I also found another few cases where the server was not replying back to the scheduler.  I think I've caught them all now.

bhroam commented:
@prakashcv13 I now have a better understanding on how the reply code works with locally generated batch requests.  The function issue_Drequest() is called with release_req.  This is the function that will be called when the request is replied to (release_req just frees the batch request).  Also in issue_Drequest(), it adds a task to call the function to the event task list.  This is the list for deferred replies.  When reply_send() is eventually called (all other reply functions call reply_send()), it will notice that the request is locally generated and move the work task from the event task list to the immediate task list.  This means it'll be called the next time through the main loop.  When it does, release_req() is called and the batch request is freed.

I'm not exactly sure why your code didn't leak the batch request structure.  You called reply_preempt_job_request() and didn't call the normal reply function.  You pulled data from the batch request in reply_preempt_job_request(), but didn't free the it there.  I don't understand why it didn't get dropped on the floor.  It obviously didn't since our weekly valgrind leak detection run would have caught the leak.

In any case, I am comfortable with how it works now and why both reply_preempt_jobs_request() and reply_send() both need to be called.

prakashcv13 commented:
@bhroam - there are conflicts with your latest commit. And yes, you are right, we need to call both.

bhroam commented:
@prakashcv13 I considered rebasing yesterday, but I try not doing that in the middle of a code review.  I will rebase today and fix the conflicts.

bhroam commented:
@prakashcv13 @arungrover I fixed the conflict and added a test for preemption via qrun.  Please review the last 2 commits.

bhroam commented:
@prakashcv13 could you take another look at this PR?  I'd like to get it checked in sooner rather than later.

prakashcv13 commented:
Hey @bhroam - I was under the impression that I had already approved it. I will merge it right away.

prakashcv13 commented:
merged #1138

bhroam commented:
Thanks @prakashcv13 and @arungrover
